File: typescript/tests/payment/application/usecases/transactionAccepting.test.ts
import { beforeEach, describe, expect, it } from 'vitest';
import { generateId } from '@/common/domain/identity/entityId.js';
import { ClientReference } from '@/common/domain/primitive/clientReference.js';
import { Money } from '@/common/domain/primitive/money.js';
import { Timestamp } from '@/common/domain/primitive/timestamp.js';
import { TransactionAccepting } from '@/payment/application/usecases/transactionAccepting.js';
import { PaymentEvent } from '@/payment/domain/events/paymentEvent.js';
import type { PaymentAccepted } from '@/payment/domain/events/paymentResultEvents.js';
import {
  transactionAccepted,
  type TransactionAccepted,
} from '@/payment/domain/events/transactionAccepted.js';
import { Payment } from '@/payment/domain/payment/payment.js';
import { PaymentId } from '@/payment/domain/payment/paymentId.js';
import { PaymentStatus } from '@/payment/domain/payment/paymentStatus.js';
import { Provider } from '@/payment/domain/payment/provider.js';
import { ProviderReference } from '@/payment/domain/payment/providerReference.js';
import { TransactionId } from '@/payment/domain/payment/transactionId.js';
import { PaymentCompletion } from '@/payment/domain/policies/paymentCompletion.js';
import { SqlitePaymentRepository } from '@/payment/infrastructure/sqlitePaymentRepository.js';
import { Uuid } from '@/common/domain/primitive/uuid.js';
import { CapturingEventPublisher } from '../../../testsupport/events/capturingEventPublisher.js';
import { DatabaseSetup } from '../../../testsupport/databaseSetup.js';

describe('TransactionAccepting', () => {
  let repository: SqlitePaymentRepository;
  let publisher: CapturingEventPublisher<PaymentEvent>;
  let transactionAccepting: TransactionAccepting;

  beforeEach(() => {
    const database = DatabaseSetup.initializeInMemoryDb('payment');
    repository = new SqlitePaymentRepository(database);
    publisher = new CapturingEventPublisher();
    transactionAccepting = new TransactionAccepting(repository, publisher, new PaymentCompletion());
  });

  describe('construction', () => {
    it('rejects null repository', () => {
      expect(() => new TransactionAccepting(null as unknown as SqlitePaymentRepository, publisher, new PaymentCompletion())).toThrow();
    });

    it('rejects null publisher', () => {
      expect(() => new TransactionAccepting(repository, null as unknown as CapturingEventPublisher<PaymentEvent>, new PaymentCompletion())).toThrow();
    });

    it('rejects null policy', () => {
      expect(() => new TransactionAccepting(repository, publisher, null as unknown as PaymentCompletion)).toThrow();
    });
  });

  describe('execution', () => {
    it('rejects null event', () => {
      expect(() => transactionAccepting.on(null as unknown as TransactionAccepted)).toThrow();
    });

    it('should accept transaction', () => {
      const paymentId = seedPayment();
      const transactionId = startTransaction(paymentId, new Money(50));
      const providerAt = new Timestamp(new Date('2026-06-07T10:30:00.000Z'));

      transactionAccepting.on(transactionAccepted(paymentId, Provider.PAYPAL, transactionId, new Money(50), providerAt));

      const loaded = repository.findById(paymentId);
      expect(loaded).not.toBeNull();
      expect(loaded!.status()).toEqual(PaymentStatus.ACCEPTED);
    });

    it('should publish accepted event', () => {
      const paymentId = seedPayment();
      const transactionId = startTransaction(paymentId, new Money(50));
      const providerAt = new Timestamp(new Date('2026-06-07T10:30:00.000Z'));

      transactionAccepting.on(transactionAccepted(paymentId, Provider.PAYPAL, transactionId, new Money(50), providerAt));

      expect(publisher.events()).toHaveLength(1);
      expect(publisher.events()[0].kind).toBe('PaymentAccepted');
      const accepted = publisher.events()[0] as PaymentAccepted;
      expect(accepted.aggregateId).toEqual(paymentId);
    });

    it('should fail if payment not found', () => {
      const paymentId = new PaymentId(Uuid.generate());
      const transactionId = new TransactionId(Uuid.generate());
      const providerAt = new Timestamp(new Date('2026-06-07T10:30:00.000Z'));

      expect(() =>
        transactionAccepting.on(transactionAccepted(paymentId, Provider.PAYPAL, transactionId, new Money(10), providerAt))
      ).toThrow();
    });

    it('should fail if already accepted', () => {
      const paymentId = seedPayment();
      const payment = repository.findById(paymentId)!;
      const acceptedTransactionId = generateId((value) => new TransactionId(value));
      const startedAt = new Timestamp(new Date('2026-06-07T10:01:00.000Z'));
      const acceptedAt = new Timestamp(new Date('2026-06-07T10:30:00.000Z'));
      payment.startTransaction(
        acceptedTransactionId,
        Provider.PAYPAL,
        new ProviderReference(Uuid.generate()),
        new Money(50),
        startedAt
      );
      payment.acceptTransaction(acceptedTransactionId, acceptedAt);
      repository.save(payment);

      const transactionId = new TransactionId(Uuid.generate());
      const providerAt = new Timestamp(new Date('2026-06-07T10:30:00.000Z'));
      expect(() =>
        transactionAccepting.on(transactionAccepted(paymentId, Provider.PAYPAL, transactionId, new Money(50), providerAt))
      ).toThrow();
    });
  });

  function seedPayment(): PaymentId {
    const paymentId = generateId((value) => new PaymentId(value));
    const payment = Payment.request(
      paymentId,
      new ClientReference(Uuid.fromString(crypto.randomUUID())),
      new Money(50),
      new Timestamp(new Date('2026-06-07T10:00:00.000Z'))
    );
    repository.save(payment);
    return paymentId;
  }

  function startTransaction(paymentId: PaymentId, amount: Money): TransactionId {
    const payment = repository.findById(paymentId)!;
    const transactionId = generateId((value) => new TransactionId(value));
    const started = payment.startTransaction(
      transactionId,
      Provider.PAYPAL,
      new ProviderReference(Uuid.generate()),
      amount,
      Timestamp.now()
    );
    repository.save(payment);
    return started.transactionId;
  }
});

File: java/src/test/java/it/giannibombelli/wsc2026/payment/application/usecases/TransactionAcceptingTest.java
package it.giannibombelli.wsc2026.payment.application.usecases;

import it.giannibombelli.wsc2026.common.domain.identity.EntityId;
import it.giannibombelli.wsc2026.common.domain.primitive.ClientReference;
import it.giannibombelli.wsc2026.common.domain.primitive.Money;
import it.giannibombelli.wsc2026.common.domain.primitive.Timestamp;
import it.giannibombelli.wsc2026.payment.domain.events.PaymentEvent;
import it.giannibombelli.wsc2026.payment.domain.events.PaymentResultEvents;
import it.giannibombelli.wsc2026.payment.domain.events.TransactionAccepted;
import it.giannibombelli.wsc2026.payment.domain.events.TransactionStarted;
import it.giannibombelli.wsc2026.payment.domain.payment.Payment;
import it.giannibombelli.wsc2026.payment.domain.payment.PaymentId;
import it.giannibombelli.wsc2026.payment.domain.payment.PaymentStatus;
import it.giannibombelli.wsc2026.payment.domain.payment.Provider;
import it.giannibombelli.wsc2026.payment.domain.payment.ProviderReference;
import it.giannibombelli.wsc2026.payment.domain.payment.TransactionId;
import it.giannibombelli.wsc2026.payment.domain.policies.PaymentCompletion;
import it.giannibombelli.wsc2026.payment.domain.ports.PaymentRepository;
import it.giannibombelli.wsc2026.payment.infrastructure.SqlitePaymentRepository;
import it.giannibombelli.wsc2026.testsupport.DatabaseSetup;
import it.giannibombelli.wsc2026.testsupport.events.CapturingEventPublisher;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import javax.sql.DataSource;
import java.math.BigDecimal;
import java.time.Instant;
import java.util.Optional;
import java.util.UUID;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

class TransactionAcceptingTest {
    private PaymentRepository repository;
    private CapturingEventPublisher<PaymentEvent> publisher;
    private TransactionAccepting transactionAccepting;

    @BeforeEach
    void setUp() {
        final DataSource dataSource = DatabaseSetup.initializeInMemoryDb("payment");
        repository = new SqlitePaymentRepository(dataSource);
        publisher = new CapturingEventPublisher<>();
        transactionAccepting = new TransactionAccepting(repository, publisher, new PaymentCompletion());
    }

    @Nested
    class Construction {
        @Test
        void rejectNullRepository() {
            assertThatThrownBy(() -> new TransactionAccepting(null, publisher, new PaymentCompletion()))
                .isInstanceOf(NullPointerException.class);
        }

        @Test
        void rejectNullPublisher() {
            assertThatThrownBy(() -> new TransactionAccepting(repository, null, new PaymentCompletion()))
                .isInstanceOf(NullPointerException.class);
        }

        @Test
        void rejectNullPolicy() {
            assertThatThrownBy(() -> new TransactionAccepting(repository, publisher, null))
                .isInstanceOf(NullPointerException.class);
        }
    }

    @Nested
    class Execution {
        @Test
        void rejectNullEvent() {
            assertThatThrownBy(() -> transactionAccepting.on(null))
                .isInstanceOf(IllegalArgumentException.class);
        }

        @Test
        void shouldAcceptTransaction() {
            PaymentId paymentId = EntityId.generate(PaymentId::new);
            UUID clientReference = UUID.randomUUID();
            Money amount = new Money(new BigDecimal("50.00"));
            Instant requestedAt = Instant.parse("2026-06-07T10:00:00Z");
            Timestamp providerAt = new Timestamp(Instant.parse("2026-06-07T10:30:00Z"));
            Payment payment = Payment.request(paymentId, new ClientReference(clientReference), amount, new Timestamp(requestedAt));
            repository.save(payment);

            TransactionId transactionId = startTransaction(paymentId, amount);

            transactionAccepting.on(new TransactionAccepted(paymentId, Provider.PAYPAL, transactionId, amount, providerAt));

            Optional<Payment> loaded = repository.findById(paymentId);
            assertThat(loaded).isPresent();
            assertThat(loaded.get().status()).isEqualTo(PaymentStatus.ACCEPTED);
        }

        @Test
        void shouldPublishAcceptedEvent() {
            PaymentId paymentId = EntityId.generate(PaymentId::new);
            UUID clientReference = UUID.randomUUID();
            Money amount = new Money(new BigDecimal("50.00"));
            Instant requestedAt = Instant.parse("2026-06-07T10:00:00Z");
            Timestamp providerAt = new Timestamp(Instant.parse("2026-06-07T10:30:00Z"));
            Payment payment = Payment.request(paymentId, new ClientReference(clientReference), amount, new Timestamp(requestedAt));
            repository.save(payment);

            TransactionId transactionId = startTransaction(paymentId, amount);

            transactionAccepting.on(new TransactionAccepted(paymentId, Provider.PAYPAL, transactionId, amount, providerAt));

            assertThat(publisher.events()).hasSize(1);
            assertThat(publisher.events().get(0)).isInstanceOf(PaymentResultEvents.PaymentAccepted.class);
            PaymentResultEvents.PaymentAccepted accepted = (PaymentResultEvents.PaymentAccepted) publisher.events().get(0);
            assertThat(accepted.aggregateId()).isEqualTo(paymentId);
            assertThat(accepted.clientReference().value()).isEqualTo(clientReference);
        }

        @Test
        void shouldFailIfPaymentNotFound() {
            PaymentId paymentId = EntityId.generate(PaymentId::new);
            TransactionId transactionId = new TransactionId(UUID.randomUUID());
            Timestamp providerAt = new Timestamp(Instant.parse("2026-06-07T10:30:00Z"));

            assertThatThrownBy(() -> transactionAccepting.on(
                new TransactionAccepted(paymentId, Provider.PAYPAL, transactionId, new Money(new BigDecimal("10.00")), providerAt)))
                .isInstanceOf(IllegalArgumentException.class);
        }

        @Test
        void shouldFailIfAlreadyAccepted() {
            PaymentId paymentId = EntityId.generate(PaymentId::new);
            UUID clientReference = UUID.randomUUID();
            Money amount = new Money(new BigDecimal("50.00"));
            Instant requestedAt = Instant.parse("2026-06-07T10:00:00Z");
            Payment payment = Payment.request(paymentId, new ClientReference(clientReference), amount, new Timestamp(requestedAt));
            TransactionId acceptedTransactionId = EntityId.generate(TransactionId::new);
            payment.startTransaction(
                acceptedTransactionId,
                Provider.PAYPAL,
                new ProviderReference(UUID.randomUUID()),
                amount,
                new Timestamp(requestedAt.plusSeconds(1))
            );
            payment.acceptTransaction(acceptedTransactionId, new Timestamp(requestedAt.plusSeconds(60)));
            repository.save(payment);

            TransactionId transactionId = new TransactionId(UUID.randomUUID());
            Timestamp providerAt = new Timestamp(Instant.parse("2026-06-07T10:30:00Z"));
            assertThatThrownBy(() -> transactionAccepting.on(
                new TransactionAccepted(paymentId, Provider.PAYPAL, transactionId, amount, providerAt)))
                .isInstanceOf(IllegalStateException.class);
        }

        private TransactionId startTransaction(PaymentId paymentId, Money amount) {
            Payment payment = repository.findById(paymentId).orElseThrow();
            TransactionId transactionId = EntityId.generate(TransactionId::new);
            TransactionStarted started = payment.startTransaction(
                transactionId,
                Provider.PAYPAL,
                new ProviderReference(UUID.randomUUID()),
                amount,
                new Timestamp(Instant.now())
            );
            repository.save(payment);
            return started.transactionId();
        }
    }
}
